pkg/boot/boot.go

```diff
@@ -0,0 +1,132 @@
+package boot
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// Config contains boot entries for a single configuration file (grub, syslinux, etc.)
+type Config struct {
+	MountPath    string
+	ConfigPath   string
+	Entries      []Entry
+	DefaultEntry int
+}
+
+// EntryType dictates the method by which kexec should use to load the new kernel
+type EntryType int
+
+// EntryType can be either Elf or Multiboot
+const (
+	Elf EntryType = iota
+	Multiboot
+)
+
+// Module represents a path to a binary along with arguments for its execution
+// The path in the module is relative to the mount path
+type Module struct {
+	Path   string
+	Params string
+}
+
+func (m Module) String() string {
+	return fmt.Sprintf("|'%v' (%v)|", m.Path, m.Params)
+}
+
+// NewModule constructs a module for a boot entry
+func NewModule(path string, args []string) Module {
+	return Module{
+		Path:   path,
+		Params: strings.Join(args, " "),
+	}
+}
+
+// Entry contains the necessary info to kexec into a new kernel
+type Entry struct {
+	Name    string
+	Type    EntryType
+	Modules []Module
+}
+
+// KexecLoad calls the appropriate kexec load routines based on the type of Entry
+func (e *Entry) KexecLoad() error {
+	switch e.Type {
+	case Multiboot:
+		// TODO: implement using kexec_load syscall
+		return syscall.ENOSYS
+	case Elf:
+		// TODO: implement using kexec_file_load syscall
```

Comment:
u-root has a kexec package under pkg that currently only supports kexec_file_load - you should extend that if possible.

Reply:
yup - there will soon be a dependency here on pkg/kexec to use the existing file_load call, then the load when it is available.